File: Best-Time-to-Buy-and-Sell-Stock-III.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int minp = Integer.MAX_VALUE;
        int firstp = 0;
        int maxp = -prices[0];
        int totalp = 0;

        for (int i = 0; i < n; i++) {
            minp = Math.min(minp, prices[i]);                    
            firstp = Math.max(firstp, prices[i] - minp);          
            maxp = Math.max(maxp, firstp - prices[i]);            
            totalp = Math.max(totalp, maxp + prices[i]);          
        }

        return totalp;
    }
}

File: Best-Time-to-Buy-and-Sell-Stock-III.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(p) {
    let minp = Number.MAX_VALUE;
    let maxp = -p[0];
    let firstp = 0;
    let totalp = 0;

    for (let i = 0; i < p.length; i++) {
        minp = Math.min(minp, p[i]);
        firstp = Math.max(firstp, p[i] - minp);
        maxp = Math.max(maxp, firstp - p[i]);
        totalp = Math.max(totalp, maxp + p[i]);
    }

    return totalp;
};
